compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+            );
+
+            // We need to handle situations like this:
+            //
+            // _2 = _1 as *const u8;
+            // StorageDead(_1);
+            // StorageLive(_3);
+            // _3 = _2 as *const ();
+            //
+            // What we want to do is replace one of these assignments with
+            //
+            // _3 = _1 as *const ();
+            //
+            // But if we only replace one of the two assignments we analyze with our new
+            // statement, we will use a local outside of its liveness range.
+            // To deal with this, we remove the two original statements and all storage markers
+            // for locals in the new statement, and consider the original locations of these
+            // statements to be free slots in the block.
+            // Then we insert all our removed StorageLive statements into the first free slots,
+            // our StorageDead statements into the last slots, and our new statement somewhere in
+            // the middle.
+            // This ensures that we do not change the location of any statements that we have not
+            // optimized, which minimizes the amount of our analysis that we have invalidated.
+
+            statements[temp_loc.statement_index].make_nop();
+            let mut statement = statements[final_loc.statement_index].replace_nop();
+            statement.kind = new_statement;
+
+            let locals = find_locals(&statement);
+
+            let mut storage_live = Vec::new();
+            let mut storage_dead = Vec::new();
+            let mut slots = vec![temp_loc.statement_index, final_loc.statement_index];
+            for (s, statement) in statements.iter_mut().enumerate() {
+                match statement.kind {
+                    StatementKind::StorageLive(l) => {
+                        if locals.contains(&l) {
+                            storage_live.push(statement.replace_nop());
+                            slots.push(s);
+                        }
+                    }
+                    StatementKind::StorageDead(l) => {
+                        if locals.contains(&l) {
+                            storage_dead.push(statement.replace_nop());
+                            slots.push(s);
+                        }
+                    }
+                    _ => {}
+                }
+            }
+            slots.sort();
+
+            assert!(slots.len() >= storage_live.len() + 1 + storage_dead.len());
+            for (slot, statement) in slots.iter().zip(storage_live.into_iter().chain([statement])) {
+                assert!(matches!(statements[*slot].kind, StatementKind::Nop));
+                statements[*slot] = statement;
+            }
+            for (slot, statement) in slots.iter().rev().zip(storage_dead.into_iter().rev()) {
+                assert!(matches!(statements[*slot].kind, StatementKind::Nop));
+                statements[*slot] = statement;
+            }
```

Comment:
Could you add a comment explaining why we don't add those storage statements to `invalidated_statements`?